src/worker.cc

```diff
@@ -1233,12 +1089,93 @@ worker_t::result_t to_response(const std::string& data,
+// since metrics are important both for on- and offline processing we keep the impl here
+// running services can use it and also data ETL that lives in mjolnir can use it
+struct statsd_client_t : public Statsd::StatsdClient {
+  statsd_client_t(const boost::property_tree::ptree& conf)
+      : Statsd::StatsdClient(conf.get<std::string>("statsd.host", ""),
+                             conf.get<int>("statsd.port", 8125),
+                             conf.get<std::string>("statsd.prefix", ""),
+                             conf.get<uint64_t>("statsd.batch_size", 500),
+                             0) {
+    auto added_tags = conf.get_child_optional("statsd.tags");
+    if (added_tags) {
+      for (const auto& tag : *added_tags) {
+        tags.push_back(tag.second.data());
+      }
+    }
+  }
+  std::vector<std::string> tags;
+};
+
+service_worker_t::service_worker_t(const boost::property_tree::ptree& config)
+    : interrupt(nullptr), statsd_client(new statsd_client_t(config)) {
 }
 service_worker_t::~service_worker_t() {
 }
 void service_worker_t::set_interrupt(const std::function<void()>* interrupt_function) {
   interrupt = interrupt_function;
 }
+void service_worker_t::cleanup() {
+  // sends metrics to statsd server over udp
+  statsd_client->flush();
+}
+void service_worker_t::enqueue_statistics(Api& api) const {
+  // nothing to do without stats
+  if (!api.has_info() || api.info().statistics().empty())
+    return;
+
+  // these have been filled out as the request progressed through the system
+  for (const auto& stat : api.info().statistics()) {
+    float frequency = stat.has_frequency() ? stat.frequency() : 1.f;
+    switch (stat.type()) {
+      case count:
+        statsd_client->count(stat.key(), static_cast<int>(stat.value() + 0.5), frequency,
+                             statsd_client->tags);
+        break;
+      case gauge:
+        statsd_client->gauge(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                             statsd_client->tags);
+        break;
+      case timing:
+        statsd_client->timing(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                              statsd_client->tags);
+        break;
+      case set:
+        statsd_client->set(stat.key(), static_cast<unsigned int>(stat.value() + 0.5), frequency,
+                           statsd_client->tags);
+        break;
+    }
+  }
+
+  // before we are done with the request, if this was not an error we log it was ok
+  if (api.info().has_error() && api.info().error()) {
+    auto worker = typeid(*this) == typeid(loki::loki_worker_t)
+                      ? ".loki."
+                      : (typeid(*this) == typeid(thor::thor_worker_t) ? ".thor." : ".odin.");
+    const auto& action = Options_Action_Enum_Name(api.options().action());
+
+    statsd_client->count(action + worker + "ok", 1, 1.f, statsd_client->tags);
+  }
```

Comment:
isn't that the wrong way around? you're setting "ok" _if_ there was an error

Reply:
yeah im missing a `!`